### DISCUSSION TITLE
issue: useFieldArray replace doesn't update Controller component value

### DISCUSSION THREAD

**Discussion by ChrisDoesCode** [react-hook-form/react-hook-form#10009] [Q&A] [answered]
### Version Number

7.43

### Codesandbox/Expo snack

https://codesandbox.io/s/small-sound-1emutl?file=/src/App.tsx

### Steps to reproduce

1. Create an app that uses RHF with controlledFields via `useForm`, `useFieldArray`, `useWatch`, and `Controller` as demonstrated in the CSB.
2. Have that app backfill the fieldArray from a DB/API call inside a `useEffect` as listed via `replace(<dbAPIDataHere>)`
3. Click `append` a couple times to add items to fieldArray
4. Modify one of the `age` properties of a field/row
5. Click `getData` -> the modified field and appended fields get modified by incoming data
6. Modify(or dont) another field's `age` then click `getData`. Nothing happens.
7. Console confirms that there should be data (via the log in useEffect), but the control's val isnt replaced

### Expected behaviour

Expected behavior here is that the controlled component using the same `control` as the fieldArray would be modified by the `replace` call.

Thoughts and tests:
* I'm guessing that this doesnt work because of the `control` needing to be changed via `reset`.
* The control updates once, but then stops listening after it's been modified by replace. Does isTouched or other need to be cleared?
* Reset is great for value replace, but it fails to preserve ordering when used in the DB+ControlledFields+fieldArray (values will get copied to the wrong index and etc)
* using `useWatch` for the specific value inside the controlled component does not solve
* using `setValue` seems overcomplex when modifying multiple rows w/ multiple fields ie. ```data.forEach((d,index)=>setValue(`item.${index...n}.<prop0...n>`,d.<prop0...n>) //for each prop in field```

Work Around:
* so far, if I pass the value from the mapped `controlledField` to the Controlled component as `watchValue` and set the value of the component to `watchValue`, it changes to the correct val. Additionally, checking if !null permits reuse of component when not using the watchValue and just using user input. Ex:
```
function CustomControl({watchVal, index, control}){
   <Controller
   control={control}
   render={({field})=>(
     <TextField
       ...
       value={watchVal != null ? watchVal : field.value}
     />
}

function App(){
  ...
  return(
    {controlledFields.map((field,index)=>(<CustomControl watchVal={field.age} index={index} control={control} />
  );
}
```

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
Please keep your CSB isolated and minimal to the issue itself and remove all the side effects.

**Comment by ChrisDoesCode** [comment]
Hi Bill thanks for your response.

I apologize, I dont follow on side effects. Happy to change however necessary.  (did remove some fodder)

For context: The CSB is isolated to demonstrate error when using `useEffect` with `replace` on a custom type in TS

<details>
<summary>Screenshot of console CSB showing watch vals & backing data vals differ from control:
</summary>
<img src="https://user-images.githubusercontent.com/22903918/221072654-39f0c191-f861-4184-a879-b89cd54f06f0.png" alt="csb screenshot">
</details>

**Comment by bluebill1049** [comment]
a couple of things: 

- why it's not using the correct field.id 
```tsx
{controlledFields.map((field, index) => (
  <div key={index}>
    <span key={`${field.id}-name`}>{`${field.name}-> `}</span>
    <NiftyControl
      key={`${field.id}-val`}
      index={index}
      control={control}
    />
  </div>
))}
```

- The following code looks suspicious as well
```tsx
const controlledFields = fields.map((field, index) => ({
  ...field,
  ...watchFields[index]
}));
```

Happy to look further if you can clean up your csb, without all the other logic, just a pure component data update with replace API. Also, can you double-check if the controller is passed through the updated data or not?

**Comment by leapful** [comment, marked as answer]
@ChrisDoesCode 

The implementation in your codesandbox is a little bit overkill because the `fields` object of `useFieldArray` is being polluted  and input elements are re-rendered multiple times that might cause wrong data.

Please note `fields` object is used to store unique key and default value for item inside field array. It was not designed to store updated item value. Just use directly `replace` method and unmodified `fields` object are enough to solve the problem.

Working demo: https://codesandbox.io/s/zen-wood-xgyxtr?file=/src/App.tsx

**Reply by ChrisDoesCode** [reply]
@leapful thank you thank you for the mention on `fields` with not storing updated item val

Our `useWatch` uc is to auto-save after user input changes any of the fields. The `controlledFields` @bluebill1049 ref'd above is from the docs, but I admit I may have created an amalgamation of example snippets.

I've got a working version using your CSB, leapful, & a useEffect that reads the`useWatch`. Thank you both kindly.

Solution: Do not pass `useWatch` or `controlledFields` into control, only fields.

**Reply by ChrisDoesCode** [reply]
Quick follow up @bluebill1049 :
I was able to get our usecase to fail again as the `id` of the fields are replaced by the DB id. 
This seems to cause major issues for the RHF's ability to update. Work around was adding a `.dbId` to my type
